### DISCUSSION TITLE
Node js github actions problem

### DISCUSSION THREAD

**Discussion by ghost** [community/community#141779] [Actions] [answered]
### Select Topic Area

General

### Body

Hi everyone,

I’m trying to set up caching for my Node.js project dependencies in a GitHub Actions workflow. The idea is to speed up my build process by reusing dependencies between builds, but for some reason, the cache doesn’t seem to be working. Every time the workflow runs, it installs all dependencies from scratch, as if the cache isn’t being restored properly.

Here’s the workflow I’m using:

yaml
Copiar código
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v2

      - name: Set up Node.js
        uses: actions/setup-node@v2
        with:
          node-version: '16'

      - name: Restore node_modules from cache
        uses: actions/cache@v2
        with:
          path: node_modules
          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
          restore-keys: |
            ${{ runner.os }}-node-

      - name: Install dependencies
        run: npm ci

      - name: Build project
        run: npm run build
Problem: Every time the workflow runs, it installs the dependencies from scratch, and it doesn’t seem to use the cache. I don’t see any messages about restoring the cache, and the logs show a full install each time.

Questions:

Why isn’t the cache being restored? Is there something wrong with my path or key setup?
Are there better practices for caching Node.js dependencies in GitHub Actions?
Thanks in advance for any help!

**Comment by alexcerezo** [comment, marked as answer]
It looks like you're on the right track, but there are a few potential issues in your setup that might explain why the cache isn't being restored. Here are some tips to troubleshoot and improve your caching strategy:

Correct Cache Path: The path to your node_modules directory may be incorrect, especially if it isn’t in the root of your repository. If your package.json is in a subdirectory, make sure you adjust the path accordingly:


path: ./your-subdirectory/node_modules

Cache Key Consistency: Your cache key is based on the hash of package-lock.json. If this file changes (even slightly), a new cache will be created. This is expected behavior, but it means that even minor changes to dependencies or metadata in package-lock.json can cause the cache to be invalidated.

To help with this, make sure you’re using restore-keys effectively:

restore-keys: |
  ${{ runner.os }}-node-
This will allow GitHub to use an older cache if the exact key isn't found, potentially speeding up your installs by reusing parts of a previous cache.

Check Cache Hit in Logs: Review your workflow logs. You should see a message like:


Cache hit occurred on the primary key ...

If you don’t see this message, it means the cache isn’t being found. Double-check that the cache key is being generated correctly, and the package-lock.json file isn’t missing or being excluded in any way.

npm vs. npm ci: You are using npm ci, which removes node_modules before installing dependencies. This can interfere with cache restoration if not handled correctly. If you use npm install instead of npm ci, it may allow cached dependencies to be reused, although npm ci is generally better for clean installs in CI environments.

Try this:


run: npm install

I hope this helps! Let me know if you see improvements or if there’s anything else I can assist with.

Best,
